File: proto-pipeline-m/src/pipeline/gemini.ts
import { stat, readFile } from "node:fs/promises";
import { basename } from "node:path";
import { GEMINI_API_KEY } from "../env.ts";

const API_BASE = "https://generativelanguage.googleapis.com";

export type UploadedFile = {
  name: string;
  uri: string;
  mimeType: string;
};

export type GenerateResult<T = unknown> = {
  data: T;
  rawText: string;
  tokensIn: number;
  tokensOut: number;
  latencyMs: number;
  model: string;
};

function inferMime(path: string): string {
  const p = path.toLowerCase();
  if (p.endsWith(".mp4")) return "video/mp4";
  if (p.endsWith(".mov")) return "video/quicktime";
  if (p.endsWith(".webm")) return "video/webm";
  if (p.endsWith(".png")) return "image/png";
  if (p.endsWith(".jpg") || p.endsWith(".jpeg")) return "image/jpeg";
  if (p.endsWith(".webp")) return "image/webp";
  if (p.endsWith(".mp3")) return "audio/mpeg";
  if (p.endsWith(".ogg")) return "audio/ogg";
  if (p.endsWith(".wav")) return "audio/wav";
  return "application/octet-stream";
}

export async function uploadFile(
  filePath: string,
  displayName?: string,
): Promise<UploadedFile> {
  const size = (await stat(filePath)).size;
  const mime = inferMime(filePath);
  const name = displayName ?? basename(filePath);

  const initRes = await fetch(
    `${API_BASE}/upload/v1beta/files?key=${GEMINI_API_KEY}`,
    {
      method: "POST",
      headers: {
        "X-Goog-Upload-Protocol": "resumable",
        "X-Goog-Upload-Command": "start",
        "X-Goog-Upload-Header-Content-Length": String(size),
        "X-Goog-Upload-Header-Content-Type": mime,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ file: { display_name: name } }),
    },
  );
  if (!initRes.ok) {
    throw new Error(
      `File upload init failed ${initRes.status}: ${await initRes.text()}`,
    );
  }
  const uploadUrl = initRes.headers.get("X-Goog-Upload-URL");
  if (!uploadUrl) throw new Error("Missing X-Goog-Upload-URL header");

  const bytes = await readFile(filePath);
  const finRes = await fetch(uploadUrl, {
    method: "POST",
    headers: {
      "Content-Length": String(size),
      "X-Goog-Upload-Offset": "0",
      "X-Goog-Upload-Command": "upload, finalize",
    },
    body: bytes,
  });
  if (!finRes.ok) {
    throw new Error(
      `File upload finalize failed ${finRes.status}: ${await finRes.text()}`,
    );
  }
  const finJson = (await finRes.json()) as {
    file?: { name?: string; uri?: string; mimeType?: string };
  };
  const file = finJson.file;
  if (!file?.name || !file.uri) {
    throw new Error(`Bad file upload response: ${JSON.stringify(finJson)}`);
  }
  return { name: file.name, uri: file.uri, mimeType: file.mimeType ?? mime };
}

export async function waitUntilActive(
  fileName: string,
  timeoutMs = 180_000,
): Promise<void> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const res = await fetch(
      `${API_BASE}/v1beta/${fileName}?key=${GEMINI_API_KEY}`,
    );
    if (!res.ok) {
      throw new Error(
        `File status check failed ${res.status}: ${await res.text()}`,
      );
    }
    const j = (await res.json()) as { state?: string };
    if (j.state === "ACTIVE") return;
    if (j.state === "FAILED") throw new Error(`File processing FAILED`);
    await new Promise((r) => setTimeout(r, 2000));
  }
  throw new Error(`File ${fileName} did not become ACTIVE within ${timeoutMs}ms`);
}

export type ContentPart =
  | { text: string }
  | { fileData: { fileUri: string; mimeType: string } };

export async function generateJson<T = unknown>(
  model: string,
  systemInstruction: string,
  userParts: ContentPart[],
): Promise<GenerateResult<T>> {
  const url = `${API_BASE}/v1beta/models/${model}:generateContent?key=${GEMINI_API_KEY}`;
  const body = {
    systemInstruction: { parts: [{ text: systemInstruction }] },
    contents: [{ role: "user", parts: userParts }],
    generationConfig: {
      responseMimeType: "application/json",
      temperature: 0.4,
    },
  };
  const t0 = Date.now();
  const res = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  const latencyMs = Date.now() - t0;
  if (!res.ok) {
    throw new Error(`generateContent ${model} ${res.status}: ${await res.text()}`);
  }
  const j = (await res.json()) as {
    candidates?: Array<{ content?: { parts?: Array<{ text?: string }> } }>;
    usageMetadata?: { promptTokenCount?: number; candidatesTokenCount?: number };
  };
  const text = j.candidates?.[0]?.content?.parts?.map((p) => p.text ?? "").join("") ?? "";
  if (!text) {
    throw new Error(`Empty response from ${model}: ${JSON.stringify(j)}`);
  }
  let data: T;
  try {
    data = JSON.parse(text) as T;
  } catch (e) {
    throw new Error(
      `Non-JSON response from ${model}: ${(e as Error).message}\nRaw: ${text.slice(0, 500)}`,
    );
  }
  return {
    data,
    rawText: text,
    tokensIn: j.usageMetadata?.promptTokenCount ?? 0,
    tokensOut: j.usageMetadata?.candidatesTokenCount ?? 0,
    latencyMs,
    model,
  };
}

export const MODELS = {
  flash: "gemini-3.1-pro-preview",
  pro: "gemini-3.1-pro-preview",
} as const;

File: pipeline/gemini.py
from __future__ import annotations

import json
import time
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass
from pathlib import Path
from typing import Any


class GeminiError(RuntimeError):
    """Raised when Gemini returns an API or transport error."""


@dataclass(frozen=True)
class UploadedFile:
    name: str
    uri: str
    mime_type: str
    state: str


class GeminiClient:
    def __init__(self, api_key: str, timeout_seconds: int = 180) -> None:
        self._api_key = api_key
        self._timeout_seconds = timeout_seconds

    def upload_file(
        self,
        path: Path,
        mime_type: str,
        display_name: str,
        poll_interval_seconds: float,
        poll_timeout_seconds: int,
    ) -> UploadedFile:
        upload_url = self._start_upload(path, mime_type, display_name)
        response = self._upload_bytes(upload_url, path)
        file_payload = response.get("file", {})
        uploaded = UploadedFile(
            name=file_payload["name"],
            uri=file_payload["uri"],
            mime_type=file_payload.get("mimeType", mime_type),
            state=file_payload.get("state", "PROCESSING"),
        )
        return self.wait_until_active(uploaded, poll_interval_seconds, poll_timeout_seconds)

    def wait_until_active(
        self,
        uploaded: UploadedFile,
        poll_interval_seconds: float,
        poll_timeout_seconds: int,
    ) -> UploadedFile:
        deadline = time.monotonic() + poll_timeout_seconds
        last_state = uploaded.state
        while time.monotonic() < deadline:
            payload = self._request_json("GET", self._api_url(uploaded.name))
            last_state = payload.get("state", last_state)
            if last_state == "ACTIVE":
                return UploadedFile(
                    name=payload["name"],
                    uri=payload["uri"],
                    mime_type=payload.get("mimeType", uploaded.mime_type),
                    state=last_state,
                )
            if last_state == "FAILED":
                raise GeminiError(f"Gemini file processing failed for {uploaded.name}")
            time.sleep(poll_interval_seconds)
        raise GeminiError(f"Timed out waiting for {uploaded.name}; last state={last_state}")

    def generate_json(
        self,
        model: str,
        prompt: str,
        file: UploadedFile | None = None,
        temperature: float = 0.2,
    ) -> dict[str, Any]:
        parts: list[dict[str, Any]] = [{"text": prompt}]
        if file is not None:
            parts.append({"file_data": {"mime_type": file.mime_type, "file_uri": file.uri}})
        payload = {
            "contents": [{"role": "user", "parts": parts}],
            "generationConfig": {
                "response_mime_type": "application/json",
                "temperature": temperature,
            },
        }
        model_resource = model if model.startswith("models/") else f"models/{model}"
        return self._request_json("POST", self._api_url(f"{model_resource}:generateContent"), payload)

    def _start_upload(self, path: Path, mime_type: str, display_name: str) -> str:
        headers = {
            "X-Goog-Upload-Protocol": "resumable",
            "X-Goog-Upload-Command": "start",
            "X-Goog-Upload-Header-Content-Length": str(path.stat().st_size),
            "X-Goog-Upload-Header-Content-Type": mime_type,
            "Content-Type": "application/json",
        }
        payload = {"file": {"display_name": display_name}}
        response, response_headers = self._request(
            "POST",
            self._upload_url("files"),
            payload,
            headers=headers,
        )
        upload_url = response_headers.get("x-goog-upload-url")
        if not upload_url:
            raise GeminiError(f"Upload start did not return x-goog-upload-url: {response}")
        return upload_url

    def _upload_bytes(self, upload_url: str, path: Path) -> dict[str, Any]:
        data = path.read_bytes()
        headers = {
            "Content-Length": str(len(data)),
            "X-Goog-Upload-Offset": "0",
            "X-Goog-Upload-Command": "upload, finalize",
        }
        response, _ = self._request("POST", upload_url, raw_data=data, headers=headers)
        return response

    def _api_url(self, resource: str) -> str:
        return self._with_key(f"https://generativelanguage.googleapis.com/v1beta/{resource}")

    def _upload_url(self, resource: str) -> str:
        return self._with_key(f"https://generativelanguage.googleapis.com/upload/v1beta/{resource}")

    def _with_key(self, url: str) -> str:
        return f"{url}?{urllib.parse.urlencode({'key': self._api_key})}"

    def _request_json(
        self,
        method: str,
        url: str,
        payload: dict[str, Any] | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        response, _ = self._request(method, url, payload, headers=headers)
        return response

    def _request(
        self,
        method: str,
        url: str,
        payload: dict[str, Any] | None = None,
        raw_data: bytes | None = None,
        headers: dict[str, str] | None = None,
    ) -> tuple[dict[str, Any], dict[str, str]]:
        request_headers = headers.copy() if headers else {}
        data = raw_data
        if payload is not None:
            data = json.dumps(payload).encode("utf-8")
            request_headers.setdefault("Content-Type", "application/json")
        request = urllib.request.Request(url, data=data, headers=request_headers, method=method)
        try:
            with urllib.request.urlopen(request, timeout=self._timeout_seconds) as response:
                body = response.read().decode("utf-8")
                parsed = json.loads(body) if body else {}
                return parsed, {key.lower(): value for key, value in response.headers.items()}
        except urllib.error.HTTPError as exc:
            body = exc.read().decode("utf-8", errors="replace")
            raise GeminiError(f"Gemini HTTP {exc.code}: {body}") from exc
        except urllib.error.URLError as exc:
            raise GeminiError(f"Gemini request failed: {exc}") from exc

